File: Node.py
import Tree
import Market
import math

class Node : 

    def __init__(self, underlying, tree, up = None, down = None, div = 0, proba_totale = 1) :

        self.tree = tree
        self.underlying = underlying
        self.up = up
        self.down = down
        self.Nup = None
        self.Ndown = None
        self.Nmid = None
        self.option_value = None
        self.forward_price = 0
        self.div = div
        self.proba_totale = proba_totale
        pass

    def create_brick(self, trunc, direction = "up", div = 0, epsilon=1e-8) :
        
        alpha = self.tree.alpha
        self.forward_price = self.forward() - div # on donne la valeur du forward 
        # au noeud pour y avoir acces dans le reste du code
        vf = False # bool pour savoir si on doit faire un prunning
        if self.proba_totale < epsilon: # si la proba est inferieur a notre epsilon
            vf = True # on met à vrai notre bool et on met les autres noeuds en none
            self.Ndown = None
            self.Nup = None

        if trunc :
            
            self.create_trunc(trunc, div) # on crée le tronc 
            return
            
        elif direction == "up" and self.down is not None :

            # Dans ce if on rentre si le noeud d'en bas est tronqué => 
            if self.down.Nup is None or self.down.Ndown is None :
                self.if_prun_up(alpha, div)
                return
            if vf : # si on doit tronquer le noeud actuel
                self.Nmid = self.down.Nup
                return
            
            self.move_up(div) # sinon, on fait le move up classique
            return
        
        # direction down (de haut en bas)
        elif direction == "down" and self.up is not None :
            # si le noeud du dessus est tronqué
            if self.up.Ndown is None:
                self.if_prun_down(alpha, div)
                return
            if vf : # si on doit tronquer notre noeud actuel
                self.Nmid = self.up.Ndown
                return

            # sinon on move down commme d'habitude
            self.move_down(div) 
                
            return
        pass

    def create_trunc(self, trunc, div) :
        # creation des valeurs a injecter dans notre noeud
        Smid = self.underlying * (math.exp(self.tree.market.r * self.tree.dt)) - div
        Sup = Smid * self.tree.alpha
        Sdown =  Smid/self.tree.alpha

        prev = self # on crée le variable qui contiendra le noeuf d'avant (celui actuel)
        self.Nmid = NodeTrunc(Smid, self.tree, None, None, 0,proba_totale=0, prev=prev) # creation de chaque noeud
        self.Nup = Node(Sup, self.tree, None, self.Nmid, 0,proba_totale=0)
        self.Ndown = Node(Sdown, self.tree, self.Nmid, None, 0,proba_totale=0)
        self.Nmid.up = self.Nup # liaison des noeuds
        self.Nmid.down = self.Ndown

        self.mise_a_jour_proba_tot(div) # on met a jour les probas
    
    def if_prun_up(self, alpha, div) :
        """
        fonction appliquée lorsque le noeud du dessous a été tronqué donc 
        on crée le mid et le up grace à la formule du alpha
        """
        self.Ndown = self.down.Nmid
        Smid = self.Ndown.underlying * alpha
        self.Nmid = Node(Smid, self.tree, down=self.Ndown, div = 0, proba_totale=0)
        Sup = self.Nmid.underlying * self.tree.alpha
        if self.Nup is None:
            self.Nup = Node(Sup, self.tree, down=self.Nmid, div=0, proba_totale=0)
        self.Nmid.up = self.Nup
        
        self.recentrage_esp_up() # on recentre
        self.mise_a_jour_proba_tot(div) # et on met à jour les probas

    def if_prun_down(self, alpha, div) :
        """
        fonction appliquée lorsque le noeud du dessus a été tronqué donc 
        on crée le mid et le down grace à la formule du alpha
        """
        self.Nup = self.up.Nmid
        Smid = self.Nup.underlying / alpha
        self.Nmid = Node(Smid, self.tree, up=self.Nup, div = 0, proba_totale=0)
        # Création du Ndown
        Sdown = self.Nmid.underlying / self.tree.alpha
        if self.Ndown is None:
            self.Ndown = Node(Sdown, self.tree, up=self.Nmid, div=0, proba_totale=0)
        self.Nmid.down = self.Ndown

        self.recentrage_esp_down() # recnetrage en esperance
        self.mise_a_jour_proba_tot(div) # mise à jour des porba totale

    def move_up(self, div) :
        """
        creation des noeuds enfants (next) lorsqu'on va vers le haut
        """
        self.Ndown = self.down.Nmid
        self.Nmid = self.down.Nup
        Sup = self.Nmid.underlying * self.tree.alpha # creation du up grace à la formule
        # du alpha
        self.Nup = Node(Sup, self.tree, down = self.Nmid, div = 0, proba_totale=0)
        self.Nmid.up = self.Nup

        self.mise_a_jour_proba_tot(div)
        self.recentrage_esp_up()
        pass

    def move_down(self,div) :
        """
        creation des noeuds enfants lorsqu'on va vers le bas
        """
        self.Nup = self.up.Nmid # on donne la valeur au nup
        self.Nmid = self.up.Ndown # nmid
        Sdown = self.Nmid.underlying / self.tree.alpha # et on crée le down
        self.Ndown = Node(Sdown, self.tree, up=self.Nmid, div = 0, proba_totale=0)
        self.Nmid.down = self.Ndown

        self.mise_a_jour_proba_tot(div)
        self.recentrage_esp_down()
        pass

    def mise_a_jour_proba_tot(self, div) :

        Pmid, Pup, Pdown = self.calcul_proba(div) # recupération des probas

        # Met à jour les probabilités cumulées
        self.Nup.proba_totale  += self.proba_totale * Pup # calcul et mise à jour des probas totales
        self.Nmid.proba_totale += self.proba_totale * Pmid
        self.Ndown.proba_totale += self.proba_totale * Pdown

        pass

    def recentrage_esp_up(self) :

        alpha = self.tree.alpha

        if self.forward_price > self.Nmid.underlying * (1 + alpha) / 2: # si on doit décaler vers le haut
                
                self.Ndown = self.Nmid
                self.Nmid = self.Nmid.up # on move up
                Sup = self.Nmid.underlying * self.tree.alpha
                self.Nup = Node(Sup, self.tree, down = self.Nmid, div = 0, proba_totale=0)
                self.Nmid.up = self.Nup

        if self.forward_price < self.Nmid.underlying * ((1 + 1/alpha) / 2): # si on doit decaler vers le bas
            
            self.Nup = self.Nmid # on move down
            self.Ndown = self.Ndown.down
            self.Nmid = self.Nmid.down

        pass

    def recentrage_esp_down(self) :
        """
        fonction qui sert à faire le recentrage en esperance lorsque qu'on est dans la direction down
        elle va verifier si on doit décaler vers le haut et le bas en fonction de l'esp et faire ce décalage proprement
        """
        alpha = self.tree.alpha
        
        val = self.forward_price
        if val> self.Nmid.underlying * (1 + alpha) / 2:
            
            self.Ndown = self.Nmid
            self.Nmid = self.Nmid.up
            self.Nup = self.Nup.up

        if val < self.Nmid.underlying * ((1 + 1/alpha) / 2):
            self.Nup = self.Nmid
            self.Nmid = self.Nmid.down
            Sdown = self.Nmid.underlying / self.tree.alpha
            self.Ndown = Node(Sdown, self.tree, up=self.Nmid, div = 0, proba_totale=0)
            self.Nmid.down = self.Ndown

    def forward(self) :
        esp = self.underlying * math.exp(self.tree.market.r * self.tree.dt) 
        return esp
    
    def esp(self, div = 0) : # calcul esp
        esp = self.Nmid.underlying * self.calcul_proba(div)[0] \
              + self.Nup.underlying * self.calcul_proba(div)[1] \
              + self.Ndown.underlying * self.calcul_proba(div)[2]
        # application de la formule d'esperance
        return esp
    
    def variance(self) :
        var = (self.underlying**2) \
            * math.exp(2*self.tree.market.r * self.tree.dt) \
            * (math.exp(self.tree.market.sigma**2 *self.tree.dt) - 1)
        return var

    def calcul_proba(self, div = 0) :
        """
        fonctionc alcule proba qui s'applique au cas avec et sans div : 
        elle applique les formule du cours en recuperant la avr et l'esp du noeud
        """
        v = self.variance()
        alpha = self.tree.alpha

        Smid = self.Nmid.underlying 
        esp = self.forward_price

        Pdown = (Smid**(-2) * (v + esp**(2)) - 1 - (alpha + 1)*(Smid**(-1) * esp - 1)) / ((1 - alpha)*(alpha**(-2) - 1))

        Pup = ((Smid**(-1) * esp - 1) - (alpha**(-1) - 1)*Pdown) / (alpha - 1)
        Pmid = 1 - Pup - Pdown

        if Pmid<0 or Pup<0 or Pdown<0 or Pmid>1 or Pup>1 or Pdown>1 : # verification des probas negatives
            print("Attention !!!!!!!!!!!!")
            print(Pmid, Pup, Pdown)
            print(alpha)
            print(self.Nmid.underlying, self.Nup.underlying, self.Ndown.underlying)
            
        return [Pmid, Pup, Pdown]

"""
création de la classe Node Trunc qui herite de la classe Node, qui defini en plus un noeud
precédent
"""
class NodeTrunc(Node) :
    
    def __init__(self, underlying, tree, up=None, down=None, div=0, proba_totale = 1, prev=None):
        super().__init__(underlying, tree, up, down, div, proba_totale)
        self.prev = prev

File: Node.cs
using System;
using System.Collections.Generic;
using System.ComponentModel.Design;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Xml.Linq;
using static System.Math;
using static System.Runtime.InteropServices.JavaScript.JSType;

namespace ConsoleApp1
{
    internal class Node
    {
        public Tree tree { get; set; }
        public double underlying { get; set; }
        public Node up { get; set; }
        public Node down { get; set; }
        public Node Nup { get; set; }
        public Node Ndown { get; set; }
        public Node Nmid { get; set; }
        public double? option_value { get; set; }
        public double forward_price { get; set; }
        public double div { get; set; }
        public double proba_totale { get; set; }

        public Node(double underlying, Tree tree, Node up = null, Node down = null, double div = 0, double proba_totale = 1)
        {
            this.tree = tree;
            this.underlying = underlying;
            this.up = up;
            this.down = down;
            this.Nup = null;
            this.Ndown = null;
            this.Nmid = null;
            this.option_value = null;
            this.forward_price = 0;
            this.div = div;
            this.proba_totale = proba_totale;
        }

        public void CreateBrick(bool trunc, string direction = "up", double div = 0, bool is_div = false)
        {
            double alpha = this.tree.alpha;
            double Smid = this.underlying * Exp(this.tree.market.r * this.tree.dt) - div;
            double Sup = Smid * this.tree.alpha;
            double Sdown = Smid / this.tree.alpha;

            this.forward_price = Forward() - div; // valeur du forward pour y avoir acces dans le reste du code
            
            bool prunning = false; // indicateur de prunning, false si pas besoin de l'apliquer true sinon
            if (this.proba_totale < 1e-10)
            {
                prunning = true;
                this.Ndown = null; // on donne deja les valeurs des noeuds down et up, None, dans le cas
                // ou on applique le prunning
                this.Nup = null;
            }
            if (trunc) // on crée le tronc avec les 3 noeuds associés qui serviront de base pour las uite
            {
                this.Nmid = new NodeTrunc(Smid, this.tree,  null, null, 0, 1, (NodeTrunc)this);
                this.Nup = new Node(Sup, this.tree,  null, this.Nmid, 0);
                this.Ndown = new Node(Sdown, this.tree, this.Nmid, null, 0);
                this.Nmid.up = this.Nup;
                this.Nmid.down = this.Ndown;
            }
            else if (direction == "up" && this.down != null) // direction up, equivalent move up
            {

                if ((this.down.Nup is null) || (this.down.Ndown is null)) // si le noeud du dessous à ete tronqué
                {
                    this.Ndown = this.down.Nmid; // on fait le décalage necessaire
                    Smid = this.Ndown.underlying * this.tree.alpha; // on calcul les valeurs avec le alpha
                    this.Nmid = new Node(Smid, this.tree, null, this.Ndown, 0, 0);
                    Sup = this.Nmid.underlying * this.tree.alpha;
                    this.Nup = new Node(Sup, this.tree, null, this.Nmid, 0, 0);
                    this.Nmid.up = this.Nup; // création des liaisons
                    this.mise_a_jour_proba(div); // on met à jour les proba totale
                }
                else if(prunning)
                {
                    this.Nmid = this.down.Nup; // on donne la valeur du mid grace au down
                    return;
                }
                else
                { // sinon on fait le move up habituel
                   this.Ndown = this.down.Nmid;
                    this.Nmid = this.down.Nup;
                    Sup = this.Nmid.underlying * this.tree.alpha;
                    this.Nup = new Node(Sup, this.tree, null, this.Nmid, 0);
                    this.Nmid.up = this.Nup;
                    this.mise_a_jour_proba(div);
                }
                // recentrage en espérance
                if (this.forward_price > this.Nmid.underlying * (1 + alpha) / 2) // decalage vers le haut
                {
                    this.Ndown = this.Nmid;
                    this.Nmid = this.Nmid.up;
                    Sup = this.Nmid.underlying * this.tree.alpha;
                    this.Nup = new Node(Sup, this.tree, null, this.Nmid, 0);
                    this.Nmid.up = this.Nup;
                }
                if (this.forward_price < this.Nmid.underlying * ((1 + 1 / alpha) / 2)) // decalage vers le bas
                {
                    this.Nup = this.Nmid;
                    this.Ndown = this.Ndown.down;
                    this.Nmid = this.Nmid.down;
                }

                return;
            }
            else if (direction == "down" && this.up != null) // move down
            {
                if ((this.up.Nup is null) || (this.up.Ndown is null)) // si le noeud du dessus a ete tronqué
                {
                    this.Nup = this.up.Nmid;
                    Smid = this.Nup.underlying / this.tree.alpha;
                    this.Nmid = new Node(Smid, this.tree, this.Nup,null, 0, 0);
                    Sdown = this.Nmid.underlying / this.tree.alpha;
                    this.Ndown = new Node(Sdown, this.tree, this.Nmid, null , 0, 0);
                    this.Nmid.down = this.Ndown;
                    this.mise_a_jour_proba(div);
                }
                else if (prunning)
                {
                    this.Nmid = this.up.Ndown;
                    return;
                }
                else
                { // move down habituel
                    this.Nup = this.up.Nmid;
                    this.Nmid = this.up.Ndown;
                    Sdown = this.Nmid.underlying / this.tree.alpha;
                    this.Ndown = new Node(Sdown, this.tree, this.Nmid, null, 0);
                    this.Nmid.down = this.Ndown;
                    this.mise_a_jour_proba(div);
                }

                double val = this.forward_price;
                // recentrage en espérance
                if (val > this.Nmid.underlying * (1 + alpha) / 2) // decalage vers le haut
                {
                    this.Ndown = this.Nmid;
                    this.Nmid = this.Nmid.up;
                    this.Nup = this.Nup.up;
                }
                if (val < this.Nmid.underlying * ((1 + 1 / alpha) / 2)) // decalage vers le bas
                {
                    this.Nup = this.Nmid;
                    this.Nmid = this.Nmid.down;
                    Sdown = this.Nmid.underlying / this.tree.alpha;
                    this.Ndown = new Node(Sdown, this.tree, this.Nmid, null, 0);
                    this.Nmid.down = this.Ndown;
                }

                return;
            }
        }

        public void mise_a_jour_proba(double div = 0)
        {

            double[] probs = this.CalculProba(div); // recuperation des probas de chaque noeud
            double p_mid = probs[0];
            double p_up = probs[1];
            double p_down = probs[2];

            // met à jour les probabilités cumulées
            this.Nup.proba_totale += this.proba_totale * p_up; // calcul est mise à jour des probas totales pour chaque noeud
            this.Nmid.proba_totale += this.proba_totale * p_mid;
            this.Ndown.proba_totale += this.proba_totale * p_down;

        }

        public double Forward()
        {
            double esp = this.underlying * Exp(this.tree.market.r * this.tree.dt);
            return esp;
        }

        public double Esp(double div = 0) // application formule de l'esp usuelle
        {
            var probs = CalculProba(div);
            double esp = this.Nmid.underlying * probs[0] + this.Nup.underlying * probs[1] + this.Ndown.underlying * probs[2];
            return esp;
        }

        public double Variance()
        {
            double var = Pow(this.underlying, 2) * Exp(2 * this.tree.market.r * this.tree.dt) *
                            (Exp(Pow(this.tree.market.sigma, 2) * this.tree.dt) - 1);
            return var;
        }

        public double[] CalculProba(double div = 0)
        {
            //fonctionc alcule proba qui s'applique au cas avec et sans div : 
            //elle applique les formule du cours en recuperant la avr et l'esp du noeud
            double v = this.Variance();
            double alpha = this.tree.alpha;

            double Smid = this.Nmid.underlying;
            double esp = this.forward_price;

            double Pdown = (Pow(Smid, -2) * (v + Pow(esp, 2)) - 1 - (alpha + 1) * (Pow(Smid, -1) * esp - 1)) /
                            ((1 - alpha) * (Pow(alpha, -2) - 1));

            double Pup = ((Pow(Smid, -1) * esp - 1) - (Pow(alpha, -1) - 1) * Pdown) / (alpha - 1);
            double Pmid = 1 - Pup - Pdown;


            return new double[] { Pmid, Pup, Pdown };
        }
        
    }
}
